python/lsst/ip/isr/linearize.py

```diff
@@ -0,0 +1,228 @@
+#
+# LSST Data Management System
+# Copyright 2016 AURA/LSST.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+import abc
+from itertools import izip
+
+import numpy as np
+
+from lsst.pipe.base import Struct
+from .isrLib import applyLookupTable
+
+__all__ = ["LinearizeBase", "LinearizeLookupTable", "LinearizeSquared"]
+
+class LinearizeBase(object):
+    """Abstract base class functor for correcting non-linearity
+    """
+    __metaclass__ = abc.ABCMeta
+    LinearityType = None
+
+    @abc.abstractmethod
+    def __call__(self, image, detector, log=None):
+        """Correct non-linearity
+
+        @param[in] image  image to be corrected (an lsst.afw.image.Image)
+        @param[in] detector  detector information (an instance of lsst::afw::cameraGeom::Detector)
+        @param[in] log  logger (an lsst.pex.logging.Log), or None to disable logging;
+                    a warning is logged if amplifiers are skipped or other worrisome events occur
+
+        @return an lsst.pipe.base.Struct containing at least the following fields:
+        - numAmps number of amplifiers found
+        - numLinearized  number of amplifiers linearized
+
+        @throw RuntimeError if the linearity type is wrong
+        @throw a subclass of Exception if linearization fails for any other reason
+        """
+        pass
+
+    def checkLinearityType(self, detector):
+        """Verify that the linearity type is correct for this detector
+
+        @warning only checks the first record of the amp info catalog
+
+        @param[in] detector  detector information (an instance of lsst::afw::cameraGeom::Detector)
+
+        @throw RuntimeError if anything doesn't match
+        """
+        ampInfoType = detector.getAmpInfoCatalog()[0].getLinearityType()
+        if self.LinearityType != ampInfoType:
+            raise RuntimeError("Linearity types don't match: %s != %s" % (self.LinearityType, ampInfoType))
```

Comment:
Would an `assert` be appropriate?  Is using the wrong linearity type a programming error?